docs/explanation/locks.rst

```diff
@@ -0,0 +1,236 @@
+The counterpart to this lock is `spin_unlock_bh()`_, which releases the lock.
+
+.. note::
+
+  **User context**: refers to the execution context of a process that is currently executing a `system call`_ on behalf of a **user-space** application.
+  When the kernel is running in user context, it is processing a request initiated by a user-space application and has access to the full range of kernel services, including the ability to sleep (i.e., block) if necessary.
+
+
+spin_lock_irq()
+^^^^^^^^^^^^^^^
+
+This variation is useful for locking between `hardware interrupts`_ and softIRQs.
+When a hardware IRQ handler shares data with a softIRQ, there are two main concerns: a softIRQ can be interrupted by a hardware interrupt, and the critical region might be entered by a hardware interrupt on another CPU.
+`spin_lock_irq()`_ can be used in softIRQ routines to address this issue.
+For hardware IRQ handlers, this lock is unnecessary since softIRQs, having lower priority, cannot interrupt hardware IRQs.
+The counterpart to this lock is `spin_unlock_irq()`_, which releases the lock.
+
+
+spin_lock_irqsave()
+^^^^^^^^^^^^^^^^^^^
+
+This variant saves the interrupt state (whether interrupts were enabled or disabled) in a flags variable, which is then passed to `spin_unlock_irqrestore()`_.
+This allows the same code to be used inside a hardware IRQ handler (where interrupts are already off) and in softIRQs (where interrupt disabling is required).
+Note that softIRQs (including tasklets and timers) run on return from hardware interrupts, so `spin_lock_irq()`_ also prevents these from executing.
+In this sense, `spin_lock_irqsave()`_ is the most general and powerful locking function.
+This locking mechanism is also used when data needs to be shared between two hardware IRQ handlers, a rare but possible scenario.
+This lock is essentially a superset of all the previous spinlock primitives.
+The counterpart to this lock is `spin_unlock_irqrestore()`_, which releases the lock and clean the flags variable.
+
+Sleeping Locks (Blocking locks)
+-------------------------------
+
+Sleeping locks have the primary characteristic that tasks holding them can be put to sleep.
+Unlike spinlocks, sleeping locks (also known as blocking locks) are used in kernel regions where the process context can sleep, typically in user contexts.
+
+Mutexes
+~~~~~~~
+
+In the Linux kernel, mutexes are a specific type of locking primitive that enforces serialization on shared memory systems.
+A mutex, short for "mutual exclusion," is used to protect critical sections of code, ensuring that only one thread or process can access a shared resource at any given time.
+The primary use case for this type of lock occurs when data needs to be shared between different user contexts.
```

Comment:
The header and explanation make it sound like this section is explanining mutexes in general. But as far as I see from kernel docs `mutex` is a type of sleeping lock, alongside the rest below.

Reply:
Yeah, I've just added a better explanation for this one!